File: packages/drop-ai/android/src/main/java/com/drop/ai/DropAIPlugin.kt
package com.drop.ai

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

/**
 * DropAI — the Capacitor bridge between the React app and the native engine.
 *
 * The React layer only ever sees this clean surface (prepare, getStatus,
 * ocr, analyzeImage, getEmbedding, generateText, answerQuestion, policy,
 * model management) plus two events: `status` and `downloadProgress`.
 */
@CapacitorPlugin(name = "DropAI")
class DropAIPlugin : Plugin() {

    private val engine: DropAIEngine by lazy { DropAIEngine.init(context) }

    override fun load() {
        engine.onStatus = { status ->
            notifyListeners("status", JSObject().put("status", status))
        }
        engine.onProgress = { progress, label ->
            val data = JSObject()
            data.put("progress", progress)
            data.put("label", label)
            notifyListeners("downloadProgress", data)
        }
    }

    @PluginMethod
    fun prepare(call: PluginCall) {
        engine.prepare { ok ->
            val ret = JSObject()
            ret.put("ok", ok)
            call.resolve(ret)
        }
    }

    @PluginMethod
    fun getStatus(call: PluginCall) {
        val ret = JSObject()
        ret.put("status", engine.statusMap())
        call.resolve(ret)
    }

    @PluginMethod
    fun getEmbedding(call: PluginCall) {
        val text = call.getString("text") ?: run {
            call.reject("Missing text")
            return
        }
        val embedding = engine.embed(text)
        val ret = JSObject()
        ret.put("embedding", embedding)
        call.resolve(ret)
    }

    @PluginMethod
    fun ocr(call: PluginCall) {
        val image = call.getString("image") ?: run {
            call.resolve(null)
            return
        }
        launch {
            val result = engine.ocr(image)
            if (result == null) {
                call.resolve(null)
            } else {
                val ret = JSObject()
                ret.put("text", result["text"])
                ret.put("language", result["language"])
                call.resolve(ret)
            }
        }
    }

    @PluginMethod
    fun analyzeImage(call: PluginCall) {
        val image = call.getString("image") ?: run {
            call.resolve(null)
            return
        }
        launch {
            val result = engine.analyzeImage(image)
            val ret = JSObject()
            if (result == null) {
                ret.putNull("analysis")
            } else {
                ret.put("analysis", JSObject(result))
            }
            call.resolve(ret)
        }
    }

    @PluginMethod
    fun generateText(call: PluginCall) {
        val prompt = call.getString("prompt") ?: run {
            call.resolve(null)
            return
        }
        val contextText = call.getString("context")
        launch {
            val ret = JSObject()
            ret.put("text", engine.generateText(prompt, contextText))
            call.resolve(ret)
        }
    }

    @PluginMethod
    fun answerQuestion(call: PluginCall) {
        val question = call.getString("question") ?: run {
            call.resolve(null)
            return
        }
        val contextText = call.getString("context") ?: ""
        launch {
            val ret = JSObject()
            ret.put("answer", engine.answerQuestion(question, contextText))
            call.resolve(ret)
        }
    }

    @PluginMethod
    fun setPolicy(call: PluginCall) {
        engine.setWifiOnly(call.getBoolean("wifiOnly", true))
        val ret = JSObject()
        ret.put("ok", true)
        call.resolve(ret)
    }

    @PluginMethod
    fun getPolicy(call: PluginCall) {
        val ret = JSObject()
        ret.put("wifiOnly", engine.getWifiOnly())
        call.resolve(ret)
    }

    @PluginMethod
    fun removeModel(call: PluginCall) {
        val ok = engine.removeModel()
        val ret = JSObject()
        ret.put("ok", ok)
        call.resolve(ret)
    }

    @PluginMethod
    fun getStorageInfo(call: PluginCall) {
        val bytes = engine.storageBytes()
        if (bytes <= 0) {
            call.resolve(null)
            return
        }
        val ret = JSObject()
        ret.put("sizeBytes", bytes)
        call.resolve(ret)
    }

    private fun launch(block: suspend () -> Unit) {
        engine.launchOn(block)
    }
}

File: packages/drop-ai/ios/Sources/DropAI/DropAIPlugin.swift
import Capacitor

/// DropAI — Capacitor bridge between the React app and the native engine.
/// Clean surface: prepare, getStatus, ocr, analyzeImage, getEmbedding,
/// generateText, answerQuestion, policy + model management, and the
/// `status` / `downloadProgress` events.
@objc(DropAIPlugin)
public class DropAIPlugin: CAPPlugin {

    public override func load() {
        DropAIEngine.shared.onStatus = { [weak self] status in
            self?.notifyListeners("status", data: ["status": status])
        }
        DropAIEngine.shared.onProgress = { [weak self] progress, label in
            self?.notifyListeners("downloadProgress", data: ["progress": progress, "label": label])
        }
    }

    @objc func prepare(_ call: CAPPluginCall) {
        DropAIEngine.shared.prepare { ok in
            call.resolve(["ok": ok])
        }
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        call.resolve(["status": DropAIEngine.shared.statusMap()])
    }

    @objc func getEmbedding(_ call: CAPPluginCall) {
        guard let text = call.getString("text") else {
            call.reject("Missing text")
            return
        }
        call.resolve(["embedding": DropAIEngine.shared.embed(text)])
    }

    @objc func ocr(_ call: CAPPluginCall) {
        guard let image = call.getString("image") else {
            call.resolve()
            return
        }
        Task {
            let result = await DropAIEngine.shared.ocr(image)
            await MainActor.run {
                if let result {
                    call.resolve(["text": result.text, "language": result.language as Any])
                } else {
                    call.resolve()
                }
            }
        }
    }

    @objc func analyzeImage(_ call: CAPPluginCall) {
        guard let image = call.getString("image") else {
            call.resolve()
            return
        }
        Task {
            let result = await DropAIEngine.shared.analyzeImage(image)
            await MainActor.run {
                if let result {
                    call.resolve(result)
                } else {
                    call.resolve(["analysis": NSNull()])
                }
            }
        }
    }

    @objc func generateText(_ call: CAPPluginCall) {
        guard let prompt = call.getString("prompt") else {
            call.resolve()
            return
        }
        let contextText = call.getString("context")
        Task {
            let text = await DropAIEngine.shared.generateText(prompt: prompt, contextText: contextText)
            await MainActor.run { call.resolve(["text": text as Any]) }
        }
    }

    @objc func answerQuestion(_ call: CAPPluginCall) {
        guard let question = call.getString("question") else {
            call.resolve()
            return
        }
        let contextText = call.getString("context") ?? ""
        Task {
            let answer = await DropAIEngine.shared.answerQuestion(question: question, contextText: contextText)
            await MainActor.run { call.resolve(["answer": answer as Any]) }
        }
    }

    @objc func setPolicy(_ call: CAPPluginCall) {
        DropAIEngine.shared.wifiOnly = call.getBool("wifiOnly", true)
        call.resolve(["ok": true])
    }

    @objc func getPolicy(_ call: CAPPluginCall) {
        call.resolve(["wifiOnly": DropAIEngine.shared.wifiOnly])
    }

    @objc func removeModel(_ call: CAPPluginCall) {
        call.resolve(["ok": DropAIEngine.shared.removeModel()])
    }

    @objc func getStorageInfo(_ call: CAPPluginCall) {
        let bytes = DropAIEngine.shared.storageBytes
        if bytes <= 0 {
            call.resolve()
            return
        }
        call.resolve(["sizeBytes": bytes])
    }
}
